pendulum_control/src/pendulum_controller.hpp

```diff
@@ -0,0 +1,128 @@
+// Copyright 2015 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef PENDULUM_DEMO_PENDULUM_CONTROLLER_HPP_
+#define PENDULUM_DEMO_PENDULUM_CONTROLLER_HPP_
+
+#include <chrono>
+
+#include <pendulum_msgs/msg/joint_command.hpp>
+#include <pendulum_msgs/msg/joint_state.hpp>
+
+#ifndef PI
+#define PI 3.14159265359
+#endif
+
+namespace pendulum_control
+{
+
+struct PIDProperties
+{
+  // Properties of a PID controller
+  double p = 1;
+  double i = 0;
+  double d = 0;
+  double command = PI / 2;
+};
+
+class PendulumController
+{
+public:
+  PendulumController(std::chrono::nanoseconds period, PIDProperties pid)
+  : publish_period_(period), pid_(pid),
+    command_message_(std::make_shared<pendulum_msgs::msg::JointCommand>()),
+    message_ready_(false)
+  {
+    command_message_->position = pid_.command;
+    dt_ = publish_period_.count() / (1000.0 * 1000.0 * 1000.0);
+    assert(!isnan(dt_));
```

Comment:
An `assert` is not very string here. If this is an invariant which should always be checked in order for the following code to work correctly it should *always* be checked. I would suggest throwing a `runtime_error` instead in order to also check in non-debug builds.

Same below.